File: problems/3001-3100/3023_find-pattern-in-infinite-stream-i/provided/go/oracle.go
package main

// Problem-provided oracle (InfiniteStream), Go side. Compiled beside every
// submission by the judge; never editable in the editor. Constructed
// from the case state: the recorded bit prefix as generic values, then
// the query budget.
type InfiniteStream struct {
	bits     []int
	position int
	budget   int64
}

// NewInfiniteStream builds the oracle from the case's construction values
// (the recorded prefix as one generic slice of bits) and the query
// budget.
func NewInfiniteStream(construction []any, budget int64) *InfiniteStream {
	items, ok := construction[0].([]any)
	if !ok {
		panic("InfiniteStream bits must be an array")
	}
	bits := make([]int, 0, len(items))
	for _, item := range items {
		value, ok := item.(int64)
		if !ok {
			panic("InfiniteStream bits must be integers")
		}
		bits = append(bits, int(value))
	}
	return &InfiniteStream{bits: bits, budget: budget}
}

// Next returns the next bit of the recorded prefix, in order.
func (stream *InfiniteStream) Next() int {
	if stream.budget <= 0 {
		panic("InfiniteStream query budget exhausted")
	}
	stream.budget--
	value := stream.bits[stream.position]
	stream.position++
	return value
}

File: problems/3001-3100/3023_find-pattern-in-infinite-stream-i/provided/rust/oracle.rs
// Problem-provided oracle (InfiniteStream), Rust side. Assembled into every
// submission's crate by the judge; never editable in the editor.
// Constructed from the case state: the recorded bit prefix as generic
// values, then the query budget.
#[allow(dead_code)]
pub struct InfiniteStream {
    bits: Vec<i32>,
    position: usize,
    budget: i64,
}

impl InfiniteStream {
    pub fn new(construction: &[OjValue], budget: i64) -> Self {
        let items = match construction.first() {
            Some(OjValue::Array(items)) => items.clone(),
            _ => panic!("InfiniteStream bits must be an array"),
        };
        let mut bits = Vec::with_capacity(items.len());
        for item in items {
            match item {
                OjValue::Int(v) => bits.push(v as i32),
                _ => panic!("InfiniteStream bits must be integers"),
            }
        }
        InfiniteStream {
            bits,
            position: 0,
            budget,
        }
    }

    pub fn next(&mut self) -> i32 {
        if self.budget <= 0 {
            panic!("InfiniteStream query budget exhausted");
        }
        self.budget -= 1;
        let value = self.bits[self.position];
        self.position += 1;
        value
    }
}
